sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/reflect/DoFnSignaturesTest.java

```diff
@@ -1192,6 +1199,239 @@ public void bar(
+      implements FeatureTest {
+    @ProcessElement
+    @RequiresTimeSortedInput
+    public void process(@Element KV<String, String> input) {}
+
+    @Override
+    public void test() {
+      assertThat(DoFnSignatures.isSplittable(this), SerializableMatchers.equalTo(false));
+      assertThat(DoFnSignatures.isStateful(this), SerializableMatchers.equalTo(true));
+      assertThat(DoFnSignatures.usesTimers(this), SerializableMatchers.equalTo(true));
+      assertThat(DoFnSignatures.usesState(this), SerializableMatchers.equalTo(true));
+      assertThat(DoFnSignatures.usesBagState(this), SerializableMatchers.equalTo(true));
+      assertThat(DoFnSignatures.usesMapState(this), SerializableMatchers.equalTo(false));
+      assertThat(DoFnSignatures.usesSetState(this), SerializableMatchers.equalTo(false));
+      assertThat(DoFnSignatures.usesValueState(this), SerializableMatchers.equalTo(true));
+      assertThat(DoFnSignatures.usesWatermarkHold(this), SerializableMatchers.equalTo(true));
+      assertThat(DoFnSignatures.requiresTimeSortedInput(this), SerializableMatchers.equalTo(true));
+    }
+  }
+
+  private static class Splittable extends DoFn<KV<String, Long>, String> implements FeatureTest {
+    @ProcessElement
+    public void process(ProcessContext c, RestrictionTracker<OffsetRange, ?> tracker) {}
+
+    @GetInitialRestriction
+    public OffsetRange getInitialRange(@Element KV<String, Long> element) {
+      return new OffsetRange(0L, element.getValue());
+    }
+
+    @Override
+    public void test() {
+      assertThat(DoFnSignatures.isSplittable(this), SerializableMatchers.equalTo(true));
+      assertThat(DoFnSignatures.isStateful(this), SerializableMatchers.equalTo(false));
+      assertThat(DoFnSignatures.usesTimers(this), SerializableMatchers.equalTo(false));
+      assertThat(DoFnSignatures.usesState(this), SerializableMatchers.equalTo(false));
+      assertThat(DoFnSignatures.usesBagState(this), SerializableMatchers.equalTo(false));
+      assertThat(DoFnSignatures.usesMapState(this), SerializableMatchers.equalTo(false));
+      assertThat(DoFnSignatures.usesSetState(this), SerializableMatchers.equalTo(false));
+      assertThat(DoFnSignatures.usesValueState(this), SerializableMatchers.equalTo(false));
+      assertThat(DoFnSignatures.usesWatermarkHold(this), SerializableMatchers.equalTo(false));
+      assertThat(DoFnSignatures.requiresTimeSortedInput(this), SerializableMatchers.equalTo(false));
+    }
+  }
+
+  private final List<FeatureTest> tests =
+      Lists.newArrayList(
+          new StatelessDoFn(),
+          new StatefulWithValueState(),
+          new StatefulWithTimers(),
+          new StatefulWithTimersAndValueState(),
+          new StatefulWithSetState(),
+          new StatefulWithMapState(),
+          new StatefulWithWatermarkHoldState(),
+          new RequiresTimeSortedInput(),
+          new Splittable());
+
+  @Test
+  public void testAllDoFnFeatures() {
+    tests.forEach(FeatureTest::test);
```

Comment:
Nit: there's a lot of logic in this test class. See https://testing.googleblog.com/2014/07/testing-on-toilet-dont-put-logic-in.html

It would be better to simply write out the boilerplate for each test, so each test can be read in a straight line with no context needed.

Reply:
I wanted to keep the declaration of a class and its features encapsulated in single class. Agree that this doesn't play well with the rest of the `DoFnSignaturesTest`. But because there is effort in the direction of superseding all this with "pipeline features", I think it is fine to keep is as is for now and drop it as part of the effort later.